### PR TITLE
Adding support for --resource-type and --list-resource-types

### PR DESCRIPTION
**Pull request by saurabh-hirani**
* Add --list-resource-type cmdline option and ListResourceTypes to support it
* Add IsValidResourceType cmdline option to validate input for --resource-type
* Add IsNukeable to verify if a resource-type should be nuked as per --resource-type input
* Add if conditions for IsNukeable to prevent unnecessary get of resources which are not in --resource-type list
* Added tests for the above functionality.

### CODE REVIEW

**Comment by saurabh-hirani** [comment]
Demo:

1. Current functioality

```
$ go run main.go aws --exclude-region us-east-1 


INFO[2019-07-06T23:00:45+08:00] Retrieving all active AWS resources
INFO[2019-07-06T23:00:49+08:00] Skipping region: us-east-1
INFO[2019-07-06T23:00:50+08:00] Checking region: us-west-1
.
.
INFO[2019-07-06T23:00:49+08:00] * ec2-i-xxxx-ap-south-1
INFO[2019-07-06T23:00:49+08:00] * ebs-vol-xxxxx-ap-south-1
INFO[2019-07-06T23:00:49+08:00] * ami-ami-xxxx-ap-south-1
INFO[2019-07-06T23:00:49+08:00] * snap-snap-xxxx-ap-south-1
```

**Comment by saurabh-hirani** [comment]
The current changes require multiple ```IsNukeable``` statements for each resource. If the ```AwsResources``` interface had a ```Get``` method implemented by all resources then we could loop ```GetAllResources``` in the same manner as we are looping for Nuking https://github.com/gruntwork-io/cloud-nuke/blob/master//aws/aws.go#L266 - thereby preventing code duplication. 

I wanted to bring this to the team's notice so that we can discuss on the same before deciding on whether this change makes sense.

**Comment by brikis98** [comment]
Apologies for the delay on this. We are super buried at the moment. I have this on my TODO list and *will* come back to it!

**Comment by brikis98** [comment]
@autero1 I'm still buried and have a giant backlog of PRs to review. Could you take this one and do some thinking on the UX it creates?

**Comment by autero1** [comment]
@brikis98 I'll take a look

**Comment by autero1** [comment]
Hi @saurabh-hirani 
Thanks for the PR!

I personally would prefer keeping the existing functionality for backwards compat - if no `resource-type` specified, delete all. @brikis98 @Etiene your thoughts?

The `--list-resource-types` is absolutely needed to know what resource type there are.

**Comment by autero1** [comment]
@saurabh-hirani, tests are failing with this:

``` 
# github.com/gruntwork-io/cloud-nuke/commands
commands/cli_test.go:40:2: undefined: assert.Greater
```

**Comment by saurabh-hirani** [comment]
@autero1 thanks for going through it - checking and reverting back.

**Comment by saurabh-hirani** [comment]
> I personally would prefer keeping the existing functionality for backwards compat - if no `resource-type` specified, delete all. @brikis98 @Etiene your thoughts?
> 
> The `--list-resource-types` is absolutely needed to know what resource type there are.

@autero1  - that is the case - not specifying -resource-type gives the existing functionality.

 
> ```
> # github.com/gruntwork-io/cloud-nuke/commands
> commands/cli_test.go:40:2: undefined: assert.Greater
> ```


I checked https://github.com/gruntwork-io/cloud-nuke/pull/65#issuecomment-519083998 and saw the above error. 

The error seems weired because assert.Greater is a valid call as per https://godoc.org/github.com/stretchr/testify/assert#Greater 

To dig deeper, I added circleci integration to my branch and tried to build it but couldn't do as https://circleci.com/gh/saurabh-hirani/cloud-nuke/1 because the first step is to pull the image from 087285199408.dkr.ecr.us-east-1.amazonaws.com as per https://github.com/gruntwork-io/cloud-nuke/blob/master/.circleci/config.yml#L4 for which I don't know the creds

I have removed the ```assert.Greater``` test case in my commit as per https://github.com/gruntwork-io/cloud-nuke/pull/65/commits/63d1f34afb065962d4ae02a46904fa407c5b3937 to verify if the build goes through - but the build was not triggered on the commit. Is there a way to trigger the build to verify?

Sorry I am not that well versed with the build setup and I need help to see if removing that statement helps. It shouldn't be the cause of the error because the local tests are passing as per 

```
$ go test -v
=== RUN   TestParseDuration
--- PASS: TestParseDuration (0.00s)
=== RUN   TestParseDurationInvalidFormat
--- PASS: TestParseDurationInvalidFormat (0.00s)
=== RUN   TestListResourceTypes
--- PASS: TestListResourceTypes (0.00s)
=== RUN   TestIsValidResourceType
--- PASS: TestIsValidResourceType (0.00s)
=== RUN   TestIsNukeable
--- PASS: TestIsNukeable (0.00s)
PASS
ok      github.com/gruntwork-io/cloud-nuke/commands     3.027s
```

**Comment by autero1** [comment]
Hi @saurabh-hirani I'll kick off the test again now

**Comment by autero1** [comment]
The tests passed after a couple of intermittent errors. Maybe try and bump `stretchr/testify` version in `Gopkg.toml` to not fail on the Greater test?

**Comment by saurabh-hirani** [comment]
Thanks @autero1  - traveling for the next day - don't have access to the source - will do it day after and update you.

**Comment by saurabh-hirani** [comment]
@autero1 The problem was due to a known issue in testify - https://github.com/stretchr/testify/issues/755 - the latest master has assert.Greater function defined. The latest version v1.3.0 doesn't have it as per the issue. I added a new commit which doesn't use the version but the "master" branch in Gopkg.toml. Can you kick off a build to verify? Let me know if this makes sense, I will squash the commits before merging.

**Comment by autero1** [comment]
@saurabh-hirani While pointing to `master` fixed the issue, I think it's better practice using fixed release versions of dependencies. It makes the tests reproducible and avoids accidentally pulling in breaking changes in the library. As such, I would not use the Greater -function, but look for an alternate solution.

Additionally, as there was a recent release that caused conflicts in the PR branch, could you merge master and resolve the conflicts. Thanks!

**Comment by saurabh-hirani** [comment]
@autero1 testify v1.4.0 released 40 mins ago! :) https://github.com/stretchr/testify/releases/tag/v1.4.0 - that solves both issues - we can use v1.4.0 and uses assert.Greater. Fixed the merge conflict. Should be good to merge.

**Comment by autero1** [comment]
Great! 😄 
Kicking off tests again

**Comment by autero1** [comment]
@saurabh-hirani remember to update the README with relevant instructions in `Usage`

**Comment by saurabh-hirani** [comment]
@autero1 - thanks for running the test. I have updated README with the relevant documentation.

File: README.md [hunk 1]
```diff
@@ -67,6 +67,27 @@ You can use the `--older-than` flag to only nuke resources that were created bef
...
+cloud-nuke aws --list-resource-types
+```
+
+### Terminate specific resource types
+
+If you want to target specific resource types (e.g ec2, ami, etc.) instead of all the supported resources you can
+do so by specifying them through the `--resource-type` flag:
+
+```shell
+cloud-nuke-aws --resource-type ec2 --resource-type ami
```
Comment (autero1) [inline README.md]: Extra dash, should be `cloud-nuke aws`

Comment (saurabh-hirani) [reply README.md]: Fixed.

**Comment by saurabh-hirani** [comment]
@autero1 fixed the typo a few days ago as per the feedback. Can you please merge  the PR? Thanks.

**Comment by autero1** [comment]
@saurabh-hirani Released https://github.com/gruntwork-io/cloud-nuke/releases/tag/v0.1.9
Thanks for the PR!

**Pull request merged**